keps/sig-api-machinery/1027-api-unions/README.md

```diff
@@ -193,123 +433,706 @@ each list item is made of:
 
 Conversion between OpenAPI v2 and OpenAPI v3 will preserve these fields.
 
-### Discriminator
+### Normalization and Validation
+
+#### Normalization
+
+Normalization refers to the process by which the API server attempts to understand
+and correct clients which may provide the server with conflicting or incomplete
+information about a union in update or patch requests.
+
+Issues primarily arise here because of version skew between a client and a server,
+such as when a client is unaware of new fields added to a union and thus doesn't
+know how to clear these new fields when trying to set a different field.
+
+For unions that follow this design, normalization is simple: the server should always respect the
+discriminator.
 
-For backward compatibility reasons, discriminators should be added to existing
-union structures as an optional string. This has a nice property that it's going
-to allow conflict detection when the selected union field is changed.
+This means that when the server receives an update request with a discriminator set to a
+given field, and multiple member fields are set it should clear all fields
+except the one pointed to by the discriminator _if and only if_ the
+discriminator has been modified. Having multiple fields set, and a discriminator
+not modified is invalid and caught later by the validation step (see below).
 
-We also do strongly recommend new APIs to be written with a discriminator, and
-tools (kube-builder) should probably enforce the presence of a discriminator in
-CRDs.
+For both custom resources and built-in types, we expect union normalization to be
```

Comment:
That's odd, admission mutation will see a somewhat invalid object then?

Reply:
Yea, I think one has to come first. I don't think I have a strong opinion if you think the mutating "normalization" should come before admission mutation (whichever comes second will see a somewhat invalid object, right)

Reply:
Discussed offline, we will keep this wording as is